Add blog post about common i18n mistakes (#741)
mattermost/mattermost-developer-documentation: site/content/blog/2020-12-04-common-i18n-mistakes.md

```jsx
return (
    <FormattedMessage
        id='email_sender.remaining'
        defaultMessage='{remaining, number} email(s) remaining.'
        values={{remaining}}
    />
);
```

```go
/*
    The translation file contains:
    {
        "id": "app.email_sender.remaining",
        "translation": "{{.Remaining}} email(s) remaining"
    }
*/

func logEmailsRemaining(remaining int) {
    log.Print(translateFunc("app.email_sender.remaining", map[string]interface{}{
        "Remaining": remaining,
    }))
}
```

Note that the React example specifies that `remaining` is a number since otherwise zero would be treated as an empty string because it's falsey.

As mentioned before, these are both technically correct since they add the number of remaining emails to the right place, but they both look ugly. Instead, both libraries support pluralization by passing in a count of some object which is the number of remaining emails in this case.

```jsx
return (
    <FormattedMessage
        id='email_sender.remaining'

>         defaultMessage='{remaining, number} {remaining, plural, one {email} other {emails}} remaining.'

Reviewer: not sure if you want to add this as an example in Spanish there is a plural word for `remaining`.
singular: restante
plural: restantes

Although you can translate it to something like
`{remaining, number} {remaining, plural, one {correo pendiente} other {correos pendientes}}.`

Author: Good idea. That'll also help reinforce how other languages do things differently. I considered trying to add some non-English examples myself, but my French is really rusty.

### Solution 1: React elements as values

This technique has us pass the link with its text as a value into another translation string that represents the whole sentence.

```jsx
return (
    <FormattedMessage
        id='popup.moreInformation'
        defaultMessage='{link} for more information.'
        values={{
            link: (
                <a
                    href='https://mattermost.com'
                    rel='noreferrer'
                    target='_blank'
                >
                    <FormattedMessage
                        id='popup.moreInformation.link'
                        defaultMessage='Click here'
                    />
                </a>
            ),
        }}
    />
);
```

You'll often see this used in older Mattermost code to solve this problem. It allows us to add more complex formatting within translated text without making it impossible to translate, but it does make it harder for translators to understand the meaning of the text. They need to spend time reconstructing the entire string which may be difficult. It also makes the code more complicated to read since developers have to worry about more React elements.

> ### Solution 2: FormattedHTMLMessage

Reviewer 2: @hmhealey I think this component/API was dropped altogether in v5

> ### Solution 2: FormattedHTMLMessage

Reviewer: Maybe avoid this one as it only applies to web?

---
title: "Avoiding Common Internationalization Mistakes"
slug: common-i18n-mistakes
date: 2020-03-26T12:00:00-05:00
categories:
    - "i18n"
author: Harrison Healey
github: hmhealey
community: harrison
---

Languages are complicated, and every language is complicated in different ways that can be hard to understand without learning every single one of them. Some languages form words from multiple characters while others have symbols that represent entire concepts. Some feature words without pluralization or gender and rely on context for that while others have two or even more genders for words. Some are very phonetic while others pronounce words seemingly at random (*cough* English, though *cough*).

> Thanks to a tremendous contribution long ago from then community member Elias Nahum, we have full support for translation throughout Mattermost, and thanks to our community of translators, Mattermost is usable by many people speaking a variety of different languages.

Reviewer 3: ```suggestion
Thanks to a tremendous contribution long ago from then-community member Elias Nahum, we have full support for translation throughout Mattermost, and thanks to our community of translators, Mattermost is usable by many people speaking a variety of different languages.
```

> Because we support so many languages, we have to be aware of how to keep our applications properly translatable. There's many easy mistakes you can make when writing an application in one language that can make it difficult to translate into others. We use [react-intl](https://formatjs.io/docs/react-intl/) in our client-side applications and [go-i18n](https://github.com/nicksnyder/go-i18n) on the server which both offer a variety of features to help with this, but we can still cause problems if we're not careful when using translated text. Here's a few examples of common problems that we've run into and how to solve them.

Reviewer 3: ```suggestion
Because we support so many languages, we have to be aware of how to keep our applications properly translatable. There are many easy mistakes you can make when writing an application in one language that can make it difficult to translate into others. We use [react-intl](https://formatjs.io/docs/react-intl/) in our client-side applications and [go-i18n](https://github.com/nicksnyder/go-i18n) on the server which both offer a range of features to help with this, but we can still cause problems if we're not careful when using translated text. Here are a few examples of common problems that we've run into and how to solve them.
```

## Mistake 1: Not translating something

This one is definitely the easiest to avoid, but it can still catch people who aren't used to writing applications outside their native language.

> For example, suppose someone is adding a new button to the web app using React. If they forget to make it to use react-intl, they may write something like this:

Reviewer 3: ```suggestion
For example, suppose someone is adding a new button to the web app using React. If they forget to make it to use ``react-intl``, they may write something like this:
```

Reviewer 3: Not sure if the formatting suggestion works in this instance, so feel free to ignore if not! I haven't suggested it going forward for that reason.

Author: I'll actually rewrite references to that as React Intl since that's how they write it in their docs. For go-i18n, they just refer to it as that, so I don't think it needs to use preformatted text.

Also, I didn't realize you could do inline code with more than one backquote despite all my time working with it 😛

```jsx
function MyButton() {
    return <button>{'Click me!'}</button>;
}
```

That's fine if only English-speaking people use your application, but that doesn't work for us. Instead, we can use react-intl's `FormattedMessage` component to do the translation for us.

```jsx
import {FormattedMessage} from 'react-intl';

function MyButton() {
    return (
        <button>
            <FormattedMessage
                id='my_button.label'
                defaultMessage='Click me!'
            />
        </button>
    );
}
```

The code ends up being a bit longer, but that's more than worth it for making the app translatable.

## Mistake 2: Hard coding translations

> This mistake is a bit rarer since it requires mixing translated and non-translated text, but it comes from a real-world example in Mattermost.

Reviewer 3: ```suggestion
This mistake is less common since it requires mixing translated and non-translated text, but it comes from a real-world example in Mattermost.
```

When you reply to an older message in Mattermost, we display "Commented on Billy's message:" to tell everyone that you're responding to Billy. For a while after we had translation support added, the code for that looked like this:

```jsx
let username = '@' + rootPost.username;
if (!username.endsWith('s')) {
    username += '\'s';
}

return (
    <FormattedMessage
        id='post.commentedOn'
        defaultMessage='Commented on {username} message:'
        values={{username}}
    />
)
```

> Here, we're using react-intl to add Billy's name into the text as a value, but we're making a mistake here by trying to make the username into a possessive adjective ourselves. By manually adding the "'s", we're applying English grammar regardless of the user's language setting, so this label won't make sense in other languages.

Reviewer 3: ```suggestion
Here, we're using react-intl to add Billy's name into the text as a value, but we're making a mistake by trying to make the username into a possessive adjective ourselves. By manually adding the "'s", we're applying English grammar regardless of the user's language setting, so this label won't make sense in other languages.
```

> Instead, the "'s" should be part of the translation string. That required changing our grammar rules slightly since we previously avoided adding "'s" to the end of names that ended in the letter S, but the new way is still correct by other popular grammar guidelines. Isn't English great?

Reviewer 3: ```suggestion
Instead, the "'s" should be part of the translation string. That required changing our grammar rules slightly since we previously avoided adding "'s" to the end of names that ended in the letter "s", but the new way is still correct according to other popular grammar guidelines. Isn't English great?
```

    />
);
```

## Mistake 3: Incorrect pluralization

Another common mistake is to forget that pluralization exists when counting something, so you end up with text that looks like "1 posts deleted". You can attempt to avoid that by adding (s) to the end of a word, but that doesn't look very nice, and other languages may not have anything comparable that they can do to replicate that. Thankfully, our translation libraries can help us out here.

Say you're adding a feature that sends out an email to a number of people, and you want it to report back on its progress in both the UI and in the server's logs. You might write something like this:

```jsx
return (
    <FormattedMessage
        id='email_sender.remaining'
        defaultMessage='{remaining, number} email(s) remaining.'
        values={{remaining}}
    />
);
```

```go
/*
    The translation file contains:
    {
        "id": "app.email_sender.remaining",
        "translation": "{{.Remaining}} email(s) remaining"
    }
*/

func logEmailsRemaining(remaining int) {
    log.Print(translateFunc("app.email_sender.remaining", map[string]interface{}{
        "Remaining": remaining,
    }))
}
```

> Note that the React example uses a feature of react-intl to specify that `remaining` is a number. Without it, "0 emails remaining" wouldn't be shown properly because it's a falsey value in JavaScript.

Reviewer 3: ```suggestion
Note that the React example uses a feature of react-intl to specify that `remaining` is a number. Without it, "0 emails remaining" wouldn't be shown properly because it's a false value in JavaScript.
```

Author: Falsy is actually a correct word here, although I just learned it's usually "falsy" without the e. It's a JavaScript term for things that are considered to be false like the number 0.

> That is an example of how we can tell react-intl to behave slightly differently though. If we use the `plural` modifier, we can specify different text based on the value of `remaining`. go-i18n has a similar feature, but it takes two different versions of the translated string instead of constructing a one that is more complicated.

Reviewer 3: ```suggestion
That is an example of how we can tell react-intl to behave slightly differently though. If we use the `plural` modifier, we can specify different text based on the value of `remaining`. go-i18n has a similar feature, but it takes two different versions of the translated string instead of constructing one that's more complicated.
```

Similar to the example on pluralization, we can actually pass translated text into another block of translated text as a value to construct the sentence.

```jsx
return (
    <FormattedMessage
        id='popup.moreInformation'
        defaultMessage='{link} for more information.'
        values={{
            link: (
                <a
                    href='https://mattermost.com'
                    rel='noreferrer'
                    target='_blank'
                >
                    <FormattedMessage
                        id='popup.moreInformation.link'
                        defaultMessage='Click here'
                    />
                </a>
            ),
        }}
    />
);
```

You'll often see this used in older Mattermost code when we're trying to format part of a string. It allows us to add more complex formatting within translated text without making it impossible to translate.

It does, however, make it harder for translators to understand the meaning of the text. They need to spend time reconstructing the entire sentence which may be difficult. It also makes the code more complicated to read for the same reason.

### Solution 2: FormattedHTMLMessage

> `FormattedHTMLMessage` was another way to solve this problem, but it was removed in more recent versions. It's a compoent provided by react-intl, similar to `FormattedMessage`, but it allows us to include HTML in the translation string instead of nesting translated text. Since it's been removed, it can't be used any more, but it demonstrates another problem with ran into in the past.

Reviewer 2: ```suggestion
`FormattedHTMLMessage` was another way to solve this problem, but it was removed in more recent versions. It's a component provided by react-intl, similar to `FormattedMessage`, but it allows us to include HTML in the translation string instead of nesting translated text. Since it's been removed, it can't be used any more, but it demonstrates another problem with ran into in the past.
```

> `FormattedHTMLMessage` was another way to solve this problem, but it was removed in more recent versions. It's a compoent provided by react-intl, similar to `FormattedMessage`, but it allows us to include HTML in the translation string instead of nesting translated text. Since it's been removed, it can't be used any more, but it demonstrates another problem with ran into in the past.

Reviewer 3: ```suggestion
`FormattedHTMLMessage` was another way to solve this problem, but it was removed in more recent versions. It's a component provided by react-intl, similar to `FormattedMessage`, but allows us to include HTML in the translation string instead of nesting translated text. Since it's been removed, it can't be used any more, but it demonstrates another problem we ran into in the past.
```

```jsx
// import {FormattedHTMLMessage} from 'react-intl';

return (
    <FormattedHTMLMessage
        id='popup.moreInformation'
        defaultMessage='<a href="https://mattermost.com" rel="noreferrer" target="_blank">Click here</a> for more information.'
    />
);
```

As with the previous solution, this allows translators to reconstruct the sentence however they need, but it comes with a few serious downsides such as:
1. Translators needed to understand HTML, including more complicated things like the extra parameters on the `a` tag.
2. It added the chance that an incorrect translation could break part of the application by accidentally including malformed HTML within a string.
3. It also had potential performance impacts since we were going outside of React's DOM manipulation optimizations by using raw HTML.
4. We lost the ability to wrap parts of the string in React elements which can have more functionality, such as encapsulating the additional parameters passed into `a` tag.

> 5. Since it used HTML, `FormattedHTMLMessage` wasn't available in `react-native` for use in our mobile apps.

Reviewer 3: ```suggestion
5. Since it used HTML, `FormattedHTMLMessage` wasn't available in react-native for use in our mobile apps.
```

Reviewer 3: Not sure my correction of react-native is correct - based on other instances of react mentions without formatting.

Author: I'll also change this to React Native since that's the proper name

> But despite these issues, it helped point us in the right direction since it lead to a custom component that is much nicer to work with.

Reviewer 3: ```suggestion
But despite these issues, it helped point us in the right direction since it led to a custom component that's much nicer to work with.
```

In addition to being able to render HTML in this way, we can even use it to format translated text with custom React components too. For example, if we had an `ExternalLink` component that encapsulated the extra parameters passed to the `a` tag, we could simplify this code even further.

```jsx
return (
    <FormattedMessage
        id='popup.moreInformation'
        defaultMessage='<link>Click here</link> for more information.'
        values={{
            link: (msg) => <ExternalLink href='https://mattermost.com'>{msg}</ExternalLink>,
        }}
    />
);
```

## Closing Thoughts

Hopefully, we can do our best to make Mattermost easily translatable and available in many different languages, even as Mattermost grows and the amount of text that needs to be translated increases. By avoiding these mistakes and using some of these techniques, we can help to reduce the workload of our translators and make their jobs easier.

> Special thanks again to everyone who help translate Mattermost and who make it usable by so many more people around the world. We tremendously appreciate your efforts and all the work you put in to make sure the translations stay complete and up to date.

Reviewer 3: ```suggestion
Special thanks again to everyone who helps translate Mattermost and who make it usable by so many more people around the world. We tremendously appreciate your efforts and all the work you put in to make sure the translations stay complete and up to date.
```
